pkg/workspace_test.go

```diff
@@ -1,30 +1,314 @@
 package v1
 
 import (
-	"github.com/asaskevich/govalidator"
+	"github.com/lib/pq"
+	"github.com/onepanelio/core/pkg/util"
+	"github.com/onepanelio/core/pkg/util/ptr"
 	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc/codes"
 	"testing"
+	"time"
 )
 
-func TestWorkspaceNameValidation_Regex(t *testing.T) {
-	bad := Workspace{
-		UID:  "test",
-		Name: "600s",
+func testWorkspaceStatusToFieldMapLaunching(t *testing.T) {
+	fm := workspaceStatusToFieldMap(&WorkspaceStatus{Phase: WorkspaceLaunching})
+
+	assert.Equal(t, fm["phase"], WorkspaceLaunching)
+	assert.Equal(t, fm["paused_at"], pq.NullTime{})
+
+	started := fm["started_at"].(time.Time)
+
+	assert.True(t, started.After(time.Time{}))
+}
+
+func testWorkspaceStatusToFieldMapPausing(t *testing.T) {
+	fm := workspaceStatusToFieldMap(&WorkspaceStatus{Phase: WorkspacePausing})
+
+	assert.Equal(t, fm["phase"], WorkspacePausing)
+	assert.Equal(t, fm["started_at"], pq.NullTime{})
+
+	paused := fm["paused_at"].(time.Time)
+
+	assert.True(t, paused.After(time.Time{}))
+}
+
+func testWorkspaceStatusToFieldMapUpdating(t *testing.T) {
+	fm := workspaceStatusToFieldMap(&WorkspaceStatus{Phase: WorkspaceUpdating})
+
+	assert.Equal(t, fm["phase"], WorkspaceUpdating)
+	assert.Equal(t, fm["paused_at"], pq.NullTime{})
+
+	updated := fm["updated_at"].(time.Time)
+
+	assert.True(t, updated.After(time.Time{}))
+}
+
+func testWorkspaceStatusToFieldMapTerminating(t *testing.T) {
+	fm := workspaceStatusToFieldMap(&WorkspaceStatus{Phase: WorkspaceTerminating})
+
+	assert.Equal(t, fm["phase"], WorkspaceTerminating)
+	assert.Equal(t, fm["paused_at"], pq.NullTime{})
+	assert.Equal(t, fm["started_at"], pq.NullTime{})
+
+	terminated := fm["terminated_at"].(time.Time)
+
+	assert.True(t, terminated.After(time.Time{}))
+}
+
+func Test_WorkspaceStatusToFieldMap(t *testing.T) {
+	testWorkspaceStatusToFieldMapLaunching(t)
```

Comment:
Can you elaborate on the prefixes in test function names? I see:

`test`
`Test`
`Test_`
`testClient` (No `_` here as opposed to the public funcs)
`TestClient_`

I am assuming `TestClient_` are tests for methods on instances of `DefaultTestClient()`, and methods prefixed with `testClient` are for the private methods. But then looking at this [line](https://github.com/onepanelio/core/pull/374/files#diff-fcca39e329f7cf9707b4269a7c7e40a7R266), that doesn't seem to be the case? Also should we consistently use `_` for both public and private methods?

Reply:
General form is `Test[<Struct>]_<Method>` where <struct> is optional. This format comes from Goland - if you type in func Test.... it'll provide auto complete options to this.

`TestClient_` are for methods on `Client` struct, which is a majority of the methods in `pkg`.
So we have `TestClient_<method_name>`. Methods like these are for direct tests of a method, they should completely test a method. 

To test specific cases, you have `testClient` methods. Format is `testClient<MethodName><Condition>`
e.g. ```testClientGetWorkflowTemplateDBEmpty``` tests the `GetWorkflowTemplateDB` method, when the there is no template - empty.

`TestClientGetWorkflowTemplateDB` would call this as a child method. 

```go
TestClientGetWorfklowTemplateDB(...) {
   testClientGetWorkflowTemplateDBEmpty(...)
}
```

In the case where we have two methods where the only name difference is case, like
`GetWorkflowTemplate` vs `getWorkflowTemplate` I've used the `private` word.

```testClientPrivateGetWorkflowTemplate```
vs
```testClientGetWorkflowTemplate``` which is for the exported member.